src/core/protocol-authorization.ts

```diff
@@ -109,16 +112,63 @@ export class ProtocolAuthorization {
     await ProtocolAuthorization.verifyActionCondition(tenant, incomingMessage, messageStore);
   }
 
+  /**
+   * Performs protocol-based authorization against the incoming RecordsQuery message.
+   * @throws {Error} if authorization fails.
+   */
+  public static async authorizeQuery(
+    tenant: string,
+    incomingMessage: RecordsQuery,
+    messageStore: MessageStore,
+  ): Promise<void> {
+    // validate that required properties exist in query filter
+    const { protocol, protocolPath, contextId } = incomingMessage.message.descriptor.filter;
+    if (protocolPath === undefined) {
+      throw new DwnError(
+        DwnErrorCode.ProtocolAuthorizationQueryFilterMissingRequiredProperties,
+        'Role-authorized queries must include `protocolPath` in the filter'
+      );
+    }
+
+    // fetch the protocol definition
+    const protocolDefinition = await ProtocolAuthorization.fetchProtocolDefinition(
+      tenant,
+      protocol!,
+      messageStore,
+    );
+
+    // get the rule set for the inbound message
+    const inboundMessageRuleSet = ProtocolAuthorization.getRuleSet(
+      protocolPath,
+      protocolDefinition,
+    );
+
+    // If the incoming message has `protocolRole` in the descriptor, validate the invoked role
+    await ProtocolAuthorization.verifyInvokedRole(
+      tenant,
+      incomingMessage,
+      protocol!,
+      contextId,
+      protocolDefinition,
+      messageStore,
+    );
+
+    // verify method invoked against the allowed actions
+    await ProtocolAuthorization.verifyAllowedActions(
+      incomingMessage,
+      inboundMessageRuleSet,
+      [], // ancestor chain is not relevant to queries
```

Comment:
Related to the other comment regarding if we will ever allow query from path-based actors:

Especially if that's still in consideration and assuming you are in agreement regarding moving out the above structure check, then it seems that it would be a trade off regarding efficiency (creating a new method) vs code reuse (using existing `authorize()`)?

I am not necessarily advocating one over the other, but wanted to first make sure this is correct understanding.

Reply:
@thehenrytsai and I discussed a bit offline. We decided that the code duplication in `authorizeQuery` and `authorize` is acceptable because combining them into one function is harder to read/reason about. In fact, we'll take it a step further by separating `authorize` into `authorizeRead` and `authorizeWrite`.